### PR TITLE
cleanup forked/golang/reflect

### PR DESCRIPTION
**Pull request by mikedanese**
* remove unreferenced third_party/forked/golang/reflect
* remove unused and non-obvious 'ab' == 'abc' behavior from DeepDerivative
* move DeepDerivative into conversions.DeepEqual
* hide .../apimachinery/third_party/forked/golang/reflect and conversion.Equalities internals

/kind cleanup
/sig api-machinery

```release-note
NONE
```

### CODE REVIEW

File: third_party/forked/golang/reflect/deep_equal_test.go [hunk 1]
```diff
@@ -1,137 +0,0 @@
...
-	}
-
-	for _, item := range table {
-		if e, a := item.equal, e.DeepEqual(item.a, item.b); e != a {
-			t.Errorf("Expected (%+v == %+v) == %v, but got %v", item.a, item.b, e, a)
-		}
-	}
-}
-
-func TestDerivates(t *testing.T) {
```
Comment (mikedanese) [inline third_party/forked/golang/reflect/deep_equal_test.go]: Not sure if we want to keep these tests around. I'm happy to move them.

**Comment by liggitt** [comment]
> Our customization to reflect can be implemented easily with cmp. We could also get rid of API equality all together by implementing Equals() methods on API types can just comparing with cmp.Equal()

I vaguely remember benchmarking this with @sttts and deciding not to do it based on the performance, but I can't find the comment thread with the results. @sttts, do you remember that?

**Comment by mikedanese** [comment]
> deciding not to do it based on the performance

Sounds reasonable.

**Pull request closed without merging**